src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbDataReader.cs

```diff
@@ -610,112 +610,75 @@ public override object GetValue(int i)
 	{
+		CheckState();
+		CheckPosition();
+
 		var count = Math.Min(_fields.Count, values.Length);
-		for (var i = 0; i < count; i++)
+		try
+		{
+			for (var i = 0; i < count; i++)
+			{
+				if (_command.ExpectedColumnTypes != null)
+				{
+					var type = _command.ExpectedColumnTypes.ElementAtOrDefault(i);
+					var nullableUnderlying = Nullable.GetUnderlyingType(type);
+					if (nullableUnderlying != null)
+					{
+						if (IsDBNull(i))
+						{
+							values[i] = null;
+							continue;
+						}
+						if (nullableUnderlying == typeof(bool))
+						{
+							values[i] = _row[i].GetBoolean();
+							continue;
+						}
+					}
+					if (type == typeof(bool))
+					{
+						values[i] = _row[i].GetBoolean();
+						continue;
+					}
+				}
+				values[i] = _row[i].GetValue();
+			}
+			return count;
+		}
+		catch (IscException ex)
 		{
-			values[i] = GetValue(i);
+			throw FbException.Create(ex);
 		}
-		return count;
 	}
 
 	public override T GetFieldValue<T>(int i)
 	{
-		CheckState();
-		CheckPosition();
-		CheckIndex(i);
-
-		var type = typeof(T);
-		type = Nullable.GetUnderlyingType(type) ?? type;
+		var type = Nullable.GetUnderlyingType(typeof(T)) ?? typeof(T);
 		try
 		{
-			if (type == typeof(bool))
-			{
-				return (T)(object)_row[i].GetBoolean();
-			}
-			else if (type == typeof(byte))
-			{
-				return (T)(object)_row[i].GetByte();
-			}
-			else if (type == typeof(char))
-			{
-				return (T)(object)_row[i].GetChar();
-			}
-			else if (type == typeof(Guid))
-			{
-				return (T)(object)_row[i].GetGuid();
-			}
-			else if (type == typeof(short))
-			{
-				return (T)(object)_row[i].GetInt16();
-			}
-			else if (type == typeof(int))
-			{
-				return (T)(object)_row[i].GetInt32();
-			}
-			else if (type == typeof(long))
-			{
-				return (T)(object)_row[i].GetInt64();
-			}
-			else if (type == typeof(float))
-			{
-				return (T)(object)_row[i].GetFloat();
-			}
-			else if (type == typeof(double))
-			{
-				return (T)(object)_row[i].GetDouble();
-			}
-			else if (type == typeof(string))
-			{
-				return (T)(object)_row[i].GetString();
-			}
-			else if (type == typeof(decimal))
-			{
-				return (T)(object)_row[i].GetDecimal();
-			}
-			else if (type == typeof(DateTime))
-			{
-				return (T)(object)_row[i].GetDateTime();
-			}
-			else if (type == typeof(TimeSpan))
-			{
-				return (T)(object)_row[i].GetTimeSpan();
-			}
-			else if (type == typeof(byte[]))
-			{
-				return (T)(object)_row[i].GetBinary();
-			}
-			else if (type == typeof(FbDecFloat))
-			{
-				return (T)(object)_row[i].GetDecFloat();
-			}
-			else if (type == typeof(BigInteger))
-			{
-				return (T)(object)_row[i].GetInt128();
-			}
-			else if (type == typeof(FbZonedDateTime))
-			{
-				return (T)(object)_row[i].GetZonedDateTime();
-			}
-			else if (type == typeof(FbZonedTime))
-			{
-				return (T)(object)_row[i].GetZonedTime();
-			}
+			return type == typeof(bool) ? (T)(object)_row[i].GetBoolean()
```

Comment:
Do not change the code "formatting" unless it adds some value (this doesn't).

Reply:
Ok, but you can remove the "else", because this isn't nessessary because of the return.
May be, the optimizer does it automatically.